internal/controllers/s3bucket/cleaner.go

```diff
@@ -0,0 +1,64 @@
+package s3bucket
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/opdev/subreconciler"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/snapp-incubator/s3-operator/pkg/consts"
+)
+
+// Cleanup cleans up the provisioned resources for the s3Bucket object
+func (r *Reconciler) Cleanup(ctx context.Context) (ctrl.Result, error) {
+	// Do the actual reconcile work
+	subrecs := []subreconciler.Fn{
+		r.removeOrRetainBucket,
+		r.removeCleanupFinalizer,
+	}
+	for _, subrec := range subrecs {
+		result, err := subrec(ctx)
+		if subreconciler.ShouldHaltOrRequeue(result, err) {
+			return subreconciler.Evaluate(result, err)
+		}
+	}
+
+	return subreconciler.Evaluate(subreconciler.DoNotRequeue())
+}
+
+func (r *Reconciler) removeOrRetainBucket(ctx context.Context) (*ctrl.Result, error) {
```

Comment:
Let's discuss this retain policy. Apparently, it lets the `S3Bucket` object be deleted without deleting the actual bucket in Ceph. If the user deletes the S3UserClaim object later, its owned buckets will be deleted too.
IMHO, this behavior should at least be documented and highlighted if not revised.

Reply:
We have two options:
1. Current webhook validation prevents the user from deleting the S3UserClaim object unless all S3Bucket objects are removed. We can revise the code to check the buckets in Ceph, not the Kubernetes cluster.
2. As you said, inform the user about this behavior.

Reply:
Hamed Jan I think it would be better to document the behavior and let the deletion propagate. (the second option)

This is however the opinion of someone on the cloud side with no experience in handling *user deletion* tickets. It would be better if we could understand the users' point of view

Reply:
Sure. I will document it in both readme and the userdoc.